src/Symfony/Component/Runtime/SymfonyRuntime.php

```diff
@@ -0,0 +1,208 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Runtime;
+
+use Symfony\Component\Console\Application;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\ArgvInput;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\ConsoleOutput;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Dotenv\Dotenv;
+use Symfony\Component\ErrorHandler\Debug;
+use Symfony\Component\ErrorHandler\ErrorHandler;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+use Symfony\Component\HttpKernel\TerminableInterface;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class SymfonyRuntime implements RuntimeInterface
+{
+    private $request;
+    private $input;
+    private $output;
+    private $application;
+    private $command;
+
+    public function __construct(array $options = [])
+    {
+        if (isset($_SERVER['argv']) && class_exists(ArgvInput::class)) {
+            $this->getInput();
+        }
+
+        if ($path = $options['dotenv'] ?? true) {
+            if (!class_exists(Dotenv::class)) {
+                throw new \LogicException(sprintf('You cannot use "%s" as the Dotenv component is not installed. Try running "composer require symfony/dotenv".', __CLASS__));
+            }
+
+            (new Dotenv())->bootEnv(\is_string($path) ? $options['dotenv'] : (\dirname(__DIR__, 3).'/.env'));
+        }
+
+        if (!class_exists(ErrorHandler::class)) {
+            throw new \LogicException(sprintf('You cannot use "%s" as the ErrorHandler component is not installed. Try running "composer require symfony/error-handler".', __CLASS__));
+        }
+
+        if ($_SERVER['APP_DEBUG']) {
+            umask(0000);
+            Debug::enable();
+        } else {
+            ErrorHandler::register();
+        }
+    }
+
+    public function getApp(\Closure $app): \Closure
+    {
+        $arguments = [];
+
+        try {
+            foreach ((new \ReflectionFunction($app))->getParameters() as $parameter) {
+                $arguments[] = $this->getArgument($parameter);
+            }
+        } catch (\InvalidArgumentException $e) {
+            if (!$parameter->isOptional()) {
```

Comment:
What would be the use-case for an optional parameter? As far as I can see, the list of allowed params are strictly defined in `getArgument()` method and there is no extension to resolve more at this moment.

Reply:
I don't know, and that's the point: neither should this code. Best effort, because we don't know how in which other context the closure might be used.

Reply:
Can we move the `try-catch` into the loop to solve the rest of the arguments if one of them breaks because it's optional?

Reply:
I'm not sure we can. We cannot skip any argument since their exact position matters.

Reply:
Maybe at least I think we can support nullable arguments? For optional yes - best practice is that after optional argument only optional can follow, so there is no sence in resolving them

Reply:
Well, I don't have a real use case for optional arguments, but I think this example will break because the argument resolution will stop at the first iteration:
```php
return function (Foobar $foo = null, Request $request = null) {
    return new Response('OK '.$request->getUri());
};
```
Although the first one couldn't be solved, we should be able to solve the second one.

But I still cannot see how `Foobar` could work as an optional closure argument in any context, if currently its resolution is not yet supported.

I think If we want to add more variables to this closure I can do it in the `use` clause instead, right? Maybe I'm wrong, but I think this "optional argument" feature won't work at this point.

Reply:
I don't think there is any benefit in supporting this.
At least I don't get what this would provide. I prefer simple straightforward logic with as little special cases as possible.

Reply:
@yceruto any stat analyzer will report this as possible NPE. The more proper way to declare what you need is 
```php
function (Foobar $foo = null, Request $request)
```
if you expect an object here always
but this is incorrent according to PHP docs https://www.php.net/manual/en/functions.arguments.php#example-140 (see examples 5 and 6).

Reply:
> I don't think there is any benefit in supporting this.

Yes, I agree and that was my point in the first place https://github.com/symfony/symfony/pull/36652#discussion_r423379557 and remove the `try-catch`.

Reply:
@scaytrase now I get what you meant. This argument is not optional, but it's nullable. Before supporting nullable arguments, and before supporting fallback to their default value for optional ones, I think we need some experimental proof that this would be useful.

By not supporting this at the beginning, we'll be able to gather this proof, if it exists. Doing it the other way around would prevent us from reverting to the current behavior.

That's also why I'm in favor of the current code: we can change that later. We wouldn't as easily be able to *not* support nullable/optional args in case we want to .

Reply:
@yceruto OK, so you want me to remove the try/catch. Works for me.

> there is no extension to resolve more at this moment.

what did this mean btw?

Reply:
> I think we need some experimental proof that this would be useful

Simple example for me is like (just dirty draft)
```php
case SomePackageClass:class:
  if (class_exist(SomePackageClass:class)) { return new SomePackageClass; } 
  else 
    if ($argument->isNullable()) { return null; }
    else
      throw new \LogicException('Install vendor/package to support SomePackageClass');
```

Reply:
> what did this mean btw?

An extension point (or possibility) to resolve any argument given to this closure.

Reply:
Extension point is already present - you can extend protected method `getArgument` in you runtime

Reply:
> Simple example for me is like (just dirty draft)

yep, exactly what I want to *not* add right away - that's complexity we don't have any use case for.

> An extension point (or possibility) to resolve any argument given to this closure.

Of course, there is an extension point, that's the purpose of the component: write an implementation of RuntimeInterface. Make it a decorator or a child class of SymfonyRuntime if you want to hook into it, or provide your own logic.

:)

Reply:
Sure, I was referring to the `SymfonyRuntime` class itself.

Reply:
Actually, even without using inheritance nor hooking into `getArgument`, it's still totally extensible, by using decoration and partial functions. "High-order" is really powerful :)

Reply:
> by using decoration and partial functions

Not sure what I've understand. Can you provide an example? Currently arguments are resolved and called with local bind. How can you resolve some additinal arguments using decoration?

> "High-order" is really powerful
Yeah, but in PHP currently there is no control on closure signatures, it's not a TypeScript where you can do 
```ts
function(bla: int): ({body: int} ) => bool { return () => true; } 
```
so in my mind it's kind of step back (but still good enough for infrastructure libs)